pkg/storage/bloom/v1/fuse.go

```diff
@@ -228,74 +228,145 @@ func (fq *FusedQuerier) Run() error {
 			return errors.Wrap(err, "seeking to fingerprint")
 		}
 
-		if !fq.bq.series.Next() {
+		if !fq.bq.Next() {
 			// no more series, we're done since we're iterating desired fingerprints in order
 			return nil
 		}
 
-		series := fq.bq.series.At()
+		series := fq.bq.At()
 		if series.Fingerprint != fp {
 			// fingerprint not found, can't remove chunks
-			level.Debug(fq.logger).Log("msg", "fingerprint not found", "fp", series.Fingerprint, "err", fq.bq.series.Err())
+			level.Debug(fq.logger).Log("msg", "fingerprint not found", "fp", series.Fingerprint, "err", fq.bq.Err())
 			fq.recordMissingFp(nextBatch, fp)
 			continue
 		}
 
-		// Now that we've found the series, we need to find the unpack the bloom
-		skip := fq.bq.blooms.LoadOffset(series.Offset)
+		fq.runSeries(schema, series, nextBatch)
+	}
+
+	return nil
+}
+
+func (fq *FusedQuerier) runSeries(schema Schema, series *SeriesWithOffsets, reqs []Request) {
+	// For a given chunk|series to be removed, it must fail to match all blooms.
+	// Because iterating/loading blooms can be expensive, we iterate blooms one at a time, collecting
+	// the removals (failures) for each (bloom, chunk) pair.
+	// At the end, we intersect the removals for each series to determine if it should be removed
+
+	type inputChunks struct {
+		Missing  ChunkRefs // chunks that do not exist in the blooms and cannot be queried
+		InBlooms ChunkRefs // chunks which do exist in the blooms and can be queried
+
+		found map[int]bool // map of the index in `InBlooms` to whether the chunk
+		// was found in _any_ of the blooms for the series. In order to
+		// be eligible for removal, a chunk must be found in _no_ blooms.
+	}
+
+	inputs := make([]inputChunks, 0, len(reqs))
+	for _, req := range reqs {
+		missing, inBlooms := req.Chks.Compare(series.Chunks, true)
+		inputs = append(inputs, inputChunks{
+			Missing:  missing,
+			InBlooms: inBlooms,
+
+			found: make(map[int]bool, len(inBlooms)),
+		})
+	}
+
+	for i, offset := range series.Offsets {
+		skip := fq.bq.blooms.LoadOffset(offset)
```

Comment:
Now that we enforce the maximum size of the bloom pages. Should we remove the max page size in the read path to decrease complexity?